### ISSUE TITLE
Retrieve pmid list - use history, misc updates

### ISSUE THREAD

**Issue by jvwong** [PathwayCommons/pubmed-dl#7]
###  Background

The `ncbi.py` module retrieves a list of pmids between a provided date range (`get_list_pmid`). The ids are written and read to file, 10 000  at a time

###  Issue

Suggestion is to skip the file i/o and just use the NCBI EUTILS history server to read off the results of a search in batches of 100 000.

---

#### Step 1. ESEARCH

   - Request: 
    - GET `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?restart=0&retmax=0&mindate=2021/01/01&maxdate=2021/01/29&retmode=json&rettype=uilist&db=pubmed&datetype=pdat&term=(eng[Language])+AND+(Journal+Article[Publication+Type])&usehistory=true`
  - Response
```json
{
  "header": {
    "type": "esearch",
    "version": "0.3"
  },
  "esearchresult": {
    "count": "141421",
    "retmax": "0",
    "retstart": "0",
    "querykey": "1",
    "webenv": "MCID_60181d5d817ffd691d6bfdfe",
    "idlist": [],
    "translationset": [... ]
  }
}
```

#### Step 2. EFETCH
   - Request: 
    - GET `https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?restart=0&retmax=100000&WebEnv=MCID_60181932fffbfa4b3f340182&query_key=1&db=pubmed&rettype=uilist&retmode=text`
  - Response
```text
33516166
33516165
33516164
33516163
33516162
33516161
33516160
33516159
33516157
...(99 990 more ids)
33242461
```
  - Here you can just cycle through the `retstart` value as before, but in this case you're just fetch from the NCBI servers 

####  Other misc updates
  - For adding params to URLs, easier to use some sort of string formatting/interpolation, e.g. `f"Hello, {name}! How's it {question}?"`  
  - EFETCH with `WebEnv` should use a `query_key` value provided by an ESEARCH result, not hard-code the value
  - ESEARCH should add  parameter to describe how the `mindate`/`maxdate` should be interpreted (e.g. `datetype=pdat`)

**Comment by Anwesh1** [comment]
Hey @jvwong, I'm done writing the code to grab the data of the server, it is so much quicker. However, I was trying to integrate it with the ```ncbi.py``` file but the file is unable to read the ```.env``` for some reason. I also changed the code from
```
dot_env_filepath = Path(__file__).absolute().parent.parent / ".env"
load_dotenv(dot_env_filepath)
```
to this
```
# dot_env_filepath = Path(__file__).absolute().parent.parent / ".env"
load_dotenv(".env")
```
because the ```.env``` is in the same file as the ```ncbi.py``` file. However, it is unable to read it and is raising the ```MissingSchema(error)```. Do you have an idea why this might be happening?

**Comment by jvwong** [comment]
Which code are you referring to?

**Comment by Anwesh1** [comment]
so in the ```ncbi.py``` file in the first few lines where you load the ```.env``` file, the code is 
```
dot_env_filepath = Path(__file__).absolute().parent.parent / ".env"
load_dotenv(dot_env_filepath)
```
I tried using that but I got the ```invalid URL ' ': No schema supplied``` error so I changed the code to this (because I put the ```.env``` and the ```ncbi.py``` in the same directory)
```
# dot_env_filepath = Path(__file__).absolute().parent.parent / ".env"
load_dotenv(".env")
```
This is how I was doing it before when we were testing it. For some reason, it still gives me the same error.
I'm fairly certain that it's a small/quick fix but I'm just not sure what it is.

**Comment by Anwesh1** [comment]
I can create a PR if that would be better?

**Comment by JohnGiorgi** [comment]
@Anwesh1 Probably best to create an issue, and highlight the exact lines of code causing you the problem the way we talked about in our meeting. Once you nail down the issue, you can open a PR to address it, and close the issue once the PR is merged.

No to sound like a broken record but I highly highly recommend going through the [GithHub Learning labs](https://lab.github.com/) first couple of lessons (particularly [this one](https://lab.github.com/githubtraining/introduction-to-github)) if you haven't already.

**Comment by Anwesh1** [comment]
Hey @JohnGiorgi thanks!! I actually completed that one, I will look at some more. 
I just wasn't sure if this error warranted an issue, sorry about that.
I will do that now.

**Comment by JohnGiorgi** [comment]
Cool, no need to be sorry!

You can also highlight the problematic lines in this issue if you feel that makes more sense than opening a new issue.

**Comment by Anwesh1** [comment]
Thanks, I opened the new issue already but I will keep that in mind for future problems. 

Please let me know if you have an idea why the problem could be happening.

**Comment by Anwesh1** [comment]
I have completed updating the method of PMID retrieval. Works much faster on the retrieval standpoint and integrates smoothly with the ```ncbi.py``` file. I can edit the original ```ncbi.py``` file with this updated method.
I have also attached the code below so you can check it before I commit.
```python
def get_list_pmid(start, end):
    search_url = f"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi?&mindate={start}&maxdate={end}&retmode=json&db=pubmed&datetype=pdat&term=(eng[Language])+AND+(Journal+Article[Publication+Type])&usehistory=y"
    search_r = requests.post(search_url)
    data = search_r.json()
    webenv = data["esearchresult"]['webenv']
    total = int(data["esearchresult"]['count'])
    fetch_url = f"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?&retmax=10000&query_key=1&db=pubmed&rettype=uilist&retmode=text&WebEnv={webenv}"
    print("Total records: "+str(total))
    count = 1
    store = []
    for i in range(0, total, 10000):
        this_fetch = fetch_url+"&retstart="+str(i)
        print("URL "+str(count)+": "+this_fetch)
        count += 1
        fetch_r = requests.post(this_fetch)
        for word in fetch_r.text.split():
            store.append(word)
    return store
```

**Comment by jvwong** [comment]
1. you need to actually use the `query_key` provided, not hard-coded
2. You can fetch up to 100 000 ids in each batch
3. Can't you use [`str.splitlines()`](https://docs.python.org/3/library/stdtypes.html#str.splitlines) instead of the loop; then concat with `+` operator?
3. Create a PR for code review
